### PR TITLE
Update NightConfig and cleanup config code

### LINKED ISSUE #32: Forge configs occasionally become corrupted
This is a continuation of https://github.com/MinecraftForge/MinecraftForge/issues/9122.

After performing an analysis of the code Forge uses to interact with the Night Config library, I identified a design flaw which seems to have the potential for race conditions and corruption. In `ConfigFileTypeHandler` when a config is first registered and loaded, a file watcher is [attached](https://github.com/MinecraftForge/MinecraftForge/blob/30f255d76c4c2e915414a24e6d8aed42f429817e/fmlcore/src/main/java/net/minecraftforge/fml/config/ConfigFileTypeHandler.java#L51). Shortly after that method returns [`save` is invoked on the config object](https://github.com/MinecraftForge/MinecraftForge/blob/30f255d76c4c2e915414a24e6d8aed42f429817e/fmlcore/src/main/java/net/minecraftforge/fml/config/ConfigTracker.java#L63C15-L63C15). In summary, this can lead to the following sequence of events:

1. The CommentedFileConfig instance is built and the watcher is registered (the file on disk is now being watched for changes)
2. Forge code invokes config.save() which calls into Night Config to write to the file synchronously.
3. As part of the process of saving, `Files.newOutputStream` is invoked with the config file's path, which, according to the [Javadoc](https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/nio/file/Files.html#newOutputStream(java.nio.file.Path,java.nio.file.OpenOption...)), "initially [truncates] an existing regular-file to a size of 0 if it exists." I verified using a simple test program that, at least on Linux, the `OutputStream` given by this method is *not* atomic. The file is truncated immediately upon return of the method, and writes are seen in the underlying file even before the stream is closed.
4. Because Night Config's file watcher is asynchronous, it will detect any changes to the underlying file, regardless of whether or not saving has actually completed. That means it may see the fully truncated file, or a half-written file. In this scenario, attempting to read from the incomplete file is very likely to result in random errors when the system encounters some corrupted version.

When testing on 1.16, I was able to reliably observe symptoms similar to the theory I have described above - by augmenting the Forge file watcher's `run` function with my own, I confirmed that the [`currentlyWriting`](https://github.com/TheElectronWill/night-config/blob/e8c9f74808fe412dc43464370b206a1fe893d493/core/src/main/java/com/electronwill/nightconfig/core/file/WriteSyncFileConfig.java#L29) flag in `WriteSyncFileConfig` (the config class in NC responsible for saving) was still `true` when the watcher was invoked, indicating that saving is NOT completed before the watcher is fired.

I think fully fixing this will require a fundamental rethink of how the file watcher is used in Forge, as ideally it needs to never be listening when `.save` is called.

For completeness, [here is a link to the Discord conversation](https://discord.com/channels/313125603924639766/852298000042164244/1130678626933813340) in which I first brought up this new theory.

### LINKED ISSUE #40: ModConfigEvent can cause concurrency issues in mods
Forge fires `ModConfigEvent.Loading` when a config is first loaded, and `ModConfigEvent.Reloading` when a config reloads as a result of the file watcher detecting a change. However, there is no locking done in the function that posts these events, meaning that it is possible for both `Loading` and `Reloading` to fire simultaneously.

Many mods have `ModConfigEvent` event handlers that are not protected against this, and will therefore occasionally throw CMEs. This is technically a bug in those mods but I also think the design of this event is confusing/misleading - it really does not make sense for a config to load and reload at the exact same time.

https://github.com/ExcessiveAmountsOfZombies/Croptopia/issues/517 is an example of this occuring in the wild; I patched it in ModernFix by wrapping the Forge function that posts the event with a wrapper that synchronizes on the underlying config object, ensuring that all mods will never see two config events posted concurrently.

### PR DESCRIPTION
**Pull request by Technici4n**
See https://github.com/neoforged/FancyModLoader/pull/172 for a list of changes.

Additionally:
- Generally cleanup `ModConfigSpec`: add missing `final` on fields, and add nullability annotations.
- Document how `ModConfigSpec` stores its contents so that the next person to work on it doesn't need to spend hours understanding it again.

Fixes #32
Fixes #40

### CODE REVIEW

File: src/main/java/net/neoforged/neoforge/common/ModConfigSpec.java [hunk 1]
```diff
@@ -244,25 +271,21 @@ private int correct(UnmodifiableConfig spec, CommentedConfig config, LinkedList<
...
     }
 
-    private boolean stringsMatchIgnoringNewlines(@Nullable Object obj1, @Nullable Object obj2) {
-        if (obj1 instanceof String string1 && obj2 instanceof String string2) {
-            if (string1.length() > 0 && string2.length() > 0) {
-                return string1.replaceAll("\r\n", "\n")
-                        .equals(string2.replaceAll("\r\n", "\n"));
-
-            }
+    private boolean stringsMatchIgnoringNewlines(@Nullable String string1, @Nullable String string2) {
```
Comment (Matyrobbrt) [inline src/main/java/net/neoforged/neoforge/common/ModConfigSpec.java]: Ignoring wouldn't be technically correct, I'd name it normalizingNewLines instead

File: src/main/java/net/neoforged/neoforge/common/ModConfigSpec.java [hunk 2]
```diff
@@ -244,25 +271,21 @@ private int correct(UnmodifiableConfig spec, CommentedConfig config, LinkedList<
...
+        if (string1 != null && string2 != null && string1.length() > 0 && string2.length() > 0) {
```
Comment (Matyrobbrt) [inline src/main/java/net/neoforged/neoforge/common/ModConfigSpec.java]: I'd replace the length checks with blank checks.

File: src/main/java/net/neoforged/neoforge/network/ConfigSync.java [hunk 3]
```diff
@@ -14,20 +14,18 @@
 import net.minecraft.client.Minecraft;
 import net.neoforged.fml.config.ConfigTracker;
 import net.neoforged.fml.config.ModConfig;
+import net.neoforged.fml.config.ModConfigs;
 import net.neoforged.neoforge.network.payload.ConfigFilePayload;
 import org.jetbrains.annotations.ApiStatus;
 
 @ApiStatus.Internal
 public class ConfigSync {
```
Comment (Matyrobbrt) [inline src/main/java/net/neoforged/neoforge/network/ConfigSync.java]: What's the point of this class being a singleton now?

**Review by Matyrobbrt — CHANGES_REQUESTED** [review]

File: src/main/java/net/neoforged/neoforge/common/ModConfigSpec.java (hunk 2 above)
Comment (Technici4n) [reply src/main/java/net/neoforged/neoforge/common/ModConfigSpec.java]: Should be good now.

**Review by Matyrobbrt — APPROVED** [review]

**Pull request merged**